[tb_client] Rework the client handle

batiati opened the pull request:
This is (hopefully) the last major refactor I wanted to land for `tb_client` before considering the API stable.  

The main motivation is that the `tb_client` API is *almost* thread-safe: while `submit()` can be called from multiple application threads, `deinit()` requires synchronization. Worse, calling `submit()` after `deinit()` results in UB.  

To work around this, language clients currently guard the client handle with a mutex. However, this is suboptimal since, in practice, it makes the internal lock redundant!  

### This PR:  

Sorry for the rather large PR! I tried to keep it small, but the yak-shaving was inevitable.  
Best reviewed commit by commit.  

- Reworked the client handle as an intrusive struct, `ClientInterface`, which includes a VTable with function pointers to the state-machine specific implementation and the synchronization state. Since the language client is responsible for allocating it, `deinit()` can free all resources used by the client while leaving the interface itself intact, ensuring subsequent calls are safe.  

- Split `tb_client` into the idiomatic Zig API (bringing us one step closer to a proper Zig client) and `tb_client_exports`, which contains the FFI API.

- Reworked the build script.

- Updated clients.

### Next steps:  

- Implement the Logging callback in the language clients (the `tb_register_log_callback` function is still subject to API changes).
- Make the C client a first-class citizen among other clients, with documentation, examples, pre-built binaries, and, of course, a stable API.

batiati committed:
Add libtb_client.zig

batiati committed:
Move tb_client_exports.zig

batiati committed:
[tb_client] Rework the client interface.

The `tb_client` API is *almost* thread-safe: while `submit()` can be called from multiple threads, `deinit()` requires synchronization. Worse, calling `submit()` after `deinit()` results in UB. To work around this, language clients currently need to guard the client handle with a mutex. This is suboptimal, as it makes the existing `submit()` lock redundant.

Introducing the new intrusive struct `ClientInterface`, which includes a VTable with function pointers to the state-machine-aware implementation and the synchronization state. Since the language client is responsible for allocating this memory, `deinit()` can free all resources used by the client while leaving the interface itself intact—ensuring subsequent calls remain safe.

Also:
- `tb_client` has been split into the idiomatic Zig API (bringing us one step closer to a proper Zig client) and the `tb_client_exports`, containing the FFI API.
- `tb_client.h` reviewed, removing the `_t` suffix from enums, adding the application log functions and other minor changes.

batiati committed:
[tb_client] Rework the build script.

- Introduced a dedicated entry point for exporting `libtb_client`, resolving the issue where exports had to be kept at the root.
- Removed the `libC` requirement for unit tests, also eliminating the need for conditional imports of the `vsr` module based on `if (@import("root") == @This())`.

batiati committed:
[tb_client] Update the C sample.

batiati committed:
[Vortex] Update tb_client

batiati committed:
[Dotnet] Update tb_client

batiati committed:
[Go] Update tb_client

batiati committed:
[Java] Update tb_client

batiati committed:
[Node] Update tb_client

matklad commented on src/clients/c/tb_client.h:
```diff
@@ -249,68 +263,85 @@ typedef enum TB_PACKET_STATUS {
-    void (*on_completion)(uintptr_t, tb_client_t, tb_packet_t*, uint64_t, const uint8_t*, uint32_t)
+    uintptr_t completion_ctx,
+    void (*completion_callback)(uintptr_t, tb_packet_t*, uint64_t, const uint8_t*, uint32_t)
 );
```
To get this to 90%, let's also pass in an allocator? This won't allow us to get rid of `libc` _immediately_, due to `@cImport`, but will keep API compatible with that. 

I suggest using Lua's version of allocator:


```
typedef void * (*lua_Alloc) (void *ud,
                             void *ptr,
                             size_t osize,
                             size_t nsize);
```

The `ud` would be `completion_ctx`, `osize` is old size, `nsize` is new size. 

https://www.lua.org/manual/5.3/manual.html

matklad replied:
Decided not do do this: we need mmap to init io-uring anyway, so might as well use it for allocation.

batiati committed:
[Python] Update tb_client.

batiati committed:
[Dotnet] Fix invalid `GCHandle.Free` during the finalizer.
It turns out that calling `GCHandle.Free` during an object's finalizer can cause the CLR to crash.
Replaced it with `GC.AllocateUninitializedArray`, which provides a much better solution by supporting a GCed array with pinned memory.

batiati committed:
[Java] Add file jni_thread_cleaner.zig

batiati committed:
[Java] Fix Windows JNI Tests
Split all JNI tests so they no longer depend on the `JNIClient` or the `VSR` module.

The `VSR` module can only be built in `Debug` or `ReleaseSafe`, but on Windows, JNI tests must run in `ReleaseFast` due to `JNI_CreateJavaVM` triggering Zig's panic handler when probing the stack size.

matklad commented on src/clients/go/pkg/native/tb_client.h:
```diff
@@ -226,6 +226,20 @@ typedef enum TB_QUERY_FILTER_FLAGS {
 } TB_QUERY_FILTER_FLAGS;
 
+typedef struct tb_client_t {
+    uint64_t reserved[4];
```
Here and in the packet, let's call this `opaque` rather than `reserved`. `reserved` is supposed to be zero, and this here is used, we just don't want to explain how!

matklad replied:
Would also be great to add a comment explaining that this is a pinned, unmovable, uncopyable struct.

batiati replied:
~Humm, I think the python implementation with a `dict` holding the `tb_client` can't guarantee stability. I'll confirm that.~

*EDIT:* That's fine; even if the `dict` instance moved values during resize, it holds just the reference to a `Client`.

matklad commented on src/clients/go/pkg/native/tb_client.h:
```diff
@@ -226,6 +226,20 @@ typedef enum TB_QUERY_FILTER_FLAGS {
+    void* user_data;
+    void* data;
+    uint32_t data_size;
+    uint16_t tag;
```
Maybe call this `user_tag`, to make it clear that this field is for the user?

matklad commented on src/clients/go/pkg/native/tb_client.h:
```diff
@@ -249,68 +263,85 @@ typedef enum TB_PACKET_STATUS {
-    tb_client_t* out_client,
+TB_INIT_STATUS tb_client_init(
+    tb_client_t *client_out,
     const uint8_t cluster_id[16],
```
wold be good to add a comment that this is little-endian

matklad reviewed (commented):
Makes _a lot_ of sense overall, but I still need a closer read here!

batiati committed:
Review
- Add support to comments when generating the `tb_client.h` file.
- Comments about pinned structs `tb_client_t` and `tb_packet_t`.
- Rename `packet.tag` to `packet.user_tag` and call sites.
- Rename `tb_client_t.reserved` as `opaque`.

matklad commented on src/clients/c/tb_client.h:
```diff
@@ -226,6 +226,26 @@ typedef enum TB_QUERY_FILTER_FLAGS {
+    uint16_t user_tag;
+    uint8_t operation;
+    uint8_t status;
+    uint8_t reserved[32];
```
This one also wants to be opaque?

matklad commented on src/clients/c/tb_client.h:
```diff
@@ -249,68 +269,87 @@ typedef enum TB_PACKET_STATUS {
+);
+
+// Registers or unregisters the application log callback.
+TB_REGISTER_LOG_CALLBACK_STATUS register_log_callback(
+    void (*callback)(TB_LOG_LEVEL, const uint8_t*, uint32_t),
+    bool debug
 );
```
Let's maybe leave it to the follow up, but it seems that the current logging API doesn't spark joy:

* I think we hit UB when we access `logging.callback` and `logging.debug` outside of the mutex. The `debug` should be an atomic enum, which communicates the log level, including "is custom function set?".
* This is needlessly global state, it should be per client.
* The semantics of `register` is weird. If you try to replace the function, you get an error, but if you _clear_ and than set, than that's OK.  I think it's best to configure a per-client function _once_, by passing it to the init, and leave any dynamic switching logic to user's code. 
* As a rule of thumb, `bool` should not be used at the API level. We should use `enum LogLevel` here, even if we only support two levels now. 

But, yeah, this is mostly unrelated to the refactor here.

batiati replied:
Agreed! Let me add this bullet point on "next steps", as this hook still needs to be implemented on most language clients.

matklad commented on src/tigerbeetle/libtb_client.zig:
I'd maybe keep this under `src/clients/c`? Took me a while to find where the exports happen.

batiati replied:
Not a strong opinion here, but for me, having all entry points in `src/tigerbeetle/` makes it easier to identify which artifacts the "TigerBeetle" product produces, especially since `libtb_client` isn't just a pure VSR client.

matklad commented on src/tigerbeetle/libtb_client.zig:
```diff
@@ -0,0 +1,31 @@
+//! Entry point for exporting the `tb_client` library.
+//! Used by language clients that rely on the shared or static library exposed by `tb_client.h`.
+//! For an idiomatic Zig API, use `vsr.tb_client` directly instead.
```
FWIW, I'd say that _all_ clients should go through the static or dynamic library (so, `.zig -> .h -> .zig` sandwich), just so that we only have one canonical interface. But no super strong opinion here.

batiati replied:
Not _all_ clients, for example, the Java and Node clients use the Zig idiomatic API to build their own JNI and NAPI modules.

matklad replied:
Yeah, that's exactly my point, I think Java&Node should _also_ go through the same C API as everyone else. I don't think we win _that_ much from the API being idiomatic Zig, as idiomatic C is also quite nice to use from Zig. And it is quite nice to have just one facade.

matklad commented on build.zig:
```diff
@@ -985,14 +1005,15 @@ fn build_go_client(
 
         const lib = b.addStaticLibrary(.{
             .name = "tb_client",
```
We have 8 occurrences of `.name = "tb_client",` in build.zig, so we might as well return from `build_c_client`?

```
fn build_c_client(
) []struct {
   target: []const u8, 
   static_library: std.Build.LazyPath, 
   dynamic_library: std.Build.LazyPath, 
}
```


But can leave to follow up

matklad replied:
This should also improve compile times quite a bit I think, currently we are redundantly rebuilding the same lib multiple times I think!

batiati replied:
I'll do in a follow up PR, as it may need a deeper refactor. 
As `build_c_client` builds for multiple targets, static and dynamic libraries, it's not easily reusable in build steps like `vortex` or `c_sample`.

batiati committed:
Simplify `ClientInterface` Mutex
- Instead of reserving 16 bytes for `std.Thread.Mutex`, implement a custom lock as an `extern` struct based on a Futex with a known 32-bit size.
- Vendored from `std.Thread.Mutex.FutexImpl`, the default implementation for Linux. Also ports unit tests.

matklad commented on src/clients/c/tb_client/context.zig:
```diff
@@ -669,19 +694,74 @@ pub fn ContextType(
-            return @alignCast(@fieldParentPtr("implementation", implementation));
-        }
+        /// Functions called by `ClientInterface`, which is thread-safe.
+        const vtable = ClientInterface.VTable{
```
Nit, I but I'd maybe expressed this as

```diff
diff --git a/src/clients/c/tb_client/context.zig b/src/clients/c/tb_client/context.zig
index ac929af29..56bca2124 100644
--- a/src/clients/c/tb_client/context.zig
+++ b/src/clients/c/tb_client/context.zig
@@ -250,7 +250,11 @@ pub fn ContextType(
             };
             errdefer context.client.deinit(context.allocator);
 
-            ClientInterface.init(client_out, context, &Context.vtable);
+            ClientInterface.init(client_out, context, comptime &.{
+                .submit_fn = vtable_submit_fn,
+                .completion_context_fn = vtable_completion_context_fn,
+                .deinit_fn = vtable_deinit_fn,
+            });
             context.interface = client_out;
             context.submitted = .{
                 .name = null,
@@ -288,6 +292,59 @@ pub fn ContextType(
             };
         }
 
+        fn vtable_submit_fn(context: *anyopaque, packet_extern: *Packet.Extern) void {
+            const self: *Context = @ptrCast(@alignCast(context));
+            assert(self.signal.status() == .running);
+
+            // Packet is caller-allocated to enable elastic intrusive-link-list-based
+            // memory management. However, some of Packet's fields are essentially private.
+            // Initialize them here to avoid threading default fields through FFI boundary.
+            const packet: *Packet = packet_extern.cast();
+            packet.* = .{
+                .user_data = packet_extern.user_data,
+                .operation = packet_extern.operation,
+                .data_size = packet_extern.data_size,
+                .data = packet_extern.data,
+                .user_tag = packet_extern.user_tag,
+                .status = .ok,
+                .next = null,
+                .batch_next = null,
+                .batch_tail = null,
+                .batch_size = 0,
+                .batch_allowed = false,
+                .phase = .submitted,
+            };
+
+            // Enqueue the packet and notify the IO thread to process it asynchronously.
+            self.submitted.push(packet);
+            self.signal.notify();
+        }
+
+        fn vtable_completion_context_fn(context: *anyopaque) usize {
+            const self: *Context = @ptrCast(@alignCast(context));
+            return self.completion_context;
+        }
+
+        fn vtable_deinit_fn(context: *anyopaque) void {
+            const self: *Context = @ptrCast(@alignCast(context));
+            assert(self.signal.status() == .running);
+
+            self.signal.stop();
+            self.thread.join();
+
+            assert(self.submitted.pop() == null);
+            assert(self.pending.pop() == null);
+
+            self.io.cancel_all();
+
+            self.signal.deinit();
+            self.client.deinit(self.allocator);
+            self.message_pool.deinit(self.allocator);
+            self.io.deinit();
+
+            self.allocator.destroy(self);
+        }
+
         fn tick(self: *Context) void {
             if (self.eviction_reason == null) {
                 self.client.tick();
@@ -703,66 +760,6 @@ pub fn ContextType(
             );
         }
 
-        /// Functions called by `ClientInterface`, which is thread-safe.
-        const vtable = ClientInterface.VTable{
-            .submit_fn = struct {
-                fn submit_fn(context: *anyopaque, packet_extern: *Packet.Extern) void {
-                    const self: *Context = @ptrCast(@alignCast(context));
-                    assert(self.signal.status() == .running);
-
-                    // Packet is caller-allocated to enable elastic intrusive-link-list-based
-                    // memory management. However, some of Packet's fields are essentially private.
-                    // Initialize them here to avoid threading default fields through FFI boundary.
-                    const packet: *Packet = packet_extern.cast();
-                    packet.* = .{
-                        .user_data = packet_extern.user_data,
-                        .operation = packet_extern.operation,
-                        .data_size = packet_extern.data_size,
-                        .data = packet_extern.data,
-                        .user_tag = packet_extern.user_tag,
-                        .status = .ok,
-                        .next = null,
-                        .batch_next = null,
-                        .batch_tail = null,
-                        .batch_size = 0,
-                        .batch_allowed = false,
-                        .phase = .submitted,
-                    };
-
-                    // Enqueue the packet and notify the IO thread to process it asynchronously.
-                    self.submitted.push(packet);
-                    self.signal.notify();
-                }
-            }.submit_fn,
-            .completion_context_fn = struct {
-                fn completion_context_fn(context: *anyopaque) usize {
-                    const self: *Context = @ptrCast(@alignCast(context));
-                    return self.completion_context;
-                }
-            }.completion_context_fn,
-            .deinit_fn = struct {
-                fn deinit_fn(context: *anyopaque) void {
-                    const self: *Context = @ptrCast(@alignCast(context));
-                    assert(self.signal.status() == .running);
-
-                    self.signal.stop();
-                    self.thread.join();
-
-                    assert(self.submitted.pop() == null);
-                    assert(self.pending.pop() == null);
-
-                    self.io.cancel_all();
-
-                    self.signal.deinit();
-                    self.client.deinit(self.allocator);
-                    self.message_pool.deinit(self.allocator);
-                    self.io.deinit();
-
-                    self.allocator.destroy(self);
-                }
-            }.deinit_fn,
-        };
-
         fn operation_from_int(op: u8) ?StateMachine.Operation {
             inline for (allowed_operations) |operation| {
                 if (op == @intFromEnum(operation)) {
```

just to have fewer deep indents

batiati replied:
Cool, I like the `vtable_` prefix on the function's name!

matklad commented on src/clients/c/tb_client/packet.zig:
```diff
@@ -35,33 +32,6 @@ pub const Packet = extern struct {
-            return self.fifo.count == 0;
-        }
-    };
-
```
🎉

matklad commented on src/clients/c/tb_client.h:
```diff
@@ -230,7 +230,7 @@ typedef enum TB_QUERY_FILTER_FLAGS {
 // throughout the lifetime of the client instance.
 typedef struct tb_client_t {
-    uint64_t opaque[4];
+    uint64_t opaque[3];
```
I'd maybe keep it as `4`, to have some extra bytes, and to be a power-of-two

batiati replied:
I'm wondering ... could we use this extra byte to somehow validate the struct? Like a hash or a magic number, so we return `CLIENT_INVALID` instead of SEGFAULT'ng if the user pass a garbage pointer.

matklad replied:
Brilliant! Could write `0xBEE71E` there and assert! 

We should also write it at the _end_ of `init`, so that, if init bails part-way through, but the user doesn't handle the error and tries to use it, we'll still assert!

matklad reviewed (commented):
LGTM, just some optional nits!

batiati committed:
Review: Use `vtable_` function prefix.

batiati committed:
Review
- Use the reserved bytes from `ClientInterface` to tag initialized handlers with a magic number, preventing the use of garbage pointer.
- Handle null pointers from tb_client the C API as `TB_CLIENT_INVALID`.
- Rename `Packet.Extern.reserved` to `opaque`.
